keps/sig-node/5067-pod-generation/README.md

```diff
@@ -0,0 +1,1095 @@
+sole writer of `status.observedGeneration`. 
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Custom-set `metadata.generation`
+
+Today, it is possible for a client to custom-set a pod's `metadata.generation` on creation, but once
+set, the `metadata.generation` cannot be updated. That means that there is a
+possibility for an external client to be setting `metadata.generation` on pod 
+create and depend on that fixed value somehow in its own reconciliation logic.
+
+That said, the `metadata.generation` field is described as ["set by the system and monotonically increasing"](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata), 
+and thus it should be clear enough that `metadata.generation` was not intended to be
+used in this way.
+
+#### Infinite loop caused by misbehaving mutating webhooks
+
+It is possible that today there exists mutating webhooks that overwrite a pod's
+status. These older webhooks would not know about `status.observedGeneration` and
+could be clearing it. This would cause an infinite loop of the kubelet
+attempting to update a pod's `status.observedGeneration` and the webhook
+clearing it.
+
+Status-mutating webhooks could break more pod features than just what is
+proposed in this KEP, so we will not attempt to solve this here. This risk can 
+be mitigated by improving the documentation on webhooks and how they can be 
+written to avoid these kinds of scenarios.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+#### API server and generation
+
+For a newly created pod, the API server will set `metadata.generation` to 1. For any updates
+to the [PodSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#podspec-v1-core), the API server will increment `metadata.generation` by 1. The logic to set new pods'
+`metadata.generation` to 1 and to increment `metadata.generation` on update will
+run after all mutating webhooks have finished. 
+
+The pod `metadata.generation` will also continue to be incremented on graceful delete or 
+deferred delete, just as the API server currently does for pods and other 
+objects today.
+
+Pod updates that would not result in `metadata.generation` being incremented include:
+* Changes to metadata (with the exception of `DeletionTimestamp`). This means 
```

Comment:
Can we include the field by field analysis for the pod referenced from the doc above? I think that way it will be clear to know which changes will result in generation being incremented.

Reply:
Done! Folded in the list of mutable podspec fields here to make it clear which fields result in a generation bump. I also folded in the discussion of what `observedGeneration` means for different fields in the "mutable fields analysis" section below, and linked the doc in both places so it is easy to find more details.